introducing-pgzx.mdx

```diff
@@ -0,0 +1,208 @@
+
+### Unit tests
+
+Postgres extensions are typically tested via a tool called `pg_regress`. This is supported by pgzx as well.
+
+But we also wanted to have unit tests. This is a bit tricky because the tests need to be compiled and run in the context of a Postgres instance. Otherwise, they won't be able to interact with the Postgres APIs.
+
+In order to solve this, pgzx registers a custom `run_tests` function via the Function manager. This function can be called from SQL (`SELECT run_tests();`) and it will run the unit tests. A test suite is a Zig struct for which each function whose name starts with test is a unit test. To register a test suite, you would typically do something like this:
+
+```
+comptime {
+    pgzx.testing.registerTests(@import("build_options").testfn, .{Tests});
+}
+```
+
+The [registerTests](https://github.com/xataio/pgzx/blob/9825dde752ee4ace2cbc594332eba100292aafd5/src/pgzx/testing.zig#L39) function is another example of `comptime` usage. It iterates over all the fields of a struct and generates the call to run the tests where the `run_tests()` function is called over SQL.
+
+## Why did we start pgzx
+
+At Xata, we have been working on a new Postgres project, which doesn't really have a name yet, so let's call it "Xata's take at distributed Postgres". It is only in its very early stages, however we'll be open sourcing it soon and we're planning to build it in public. It will be somewhat similar to Citus, yet with some different choices, especially around usage and DX, based on what we learned running Xata for the past couple of years.
+
+For this project, we considered two potential implementation directions:
+
+1.  as an external proxy, like Vitess does for MySQL
+2.  as a Postgres extension, like Citus
+3.  as a fork of Postgres, like Greenplum
+
+We have some experience with the first option because that's how our current Xata Proxy works, see more details in this [blog post](https://xata.io/blog/serverless-postgres-platform), so we were almost biased towards it. We knew we can speak the Postgres wire protocol, parse the incoming queries and understand them to a fairly deep level, as well as create distributed transactions via 2PC. However, we also knew that if we were able to leverage existing Postgres code it will open up a lot more options. Given the long term vision of the project, we decided to go with the second option.
+
+The next node in our decision tree was the programming language. The main options we considered were C, Rust, and Zig. While there are pros and cons to each of these options, which we might cover in a future blog post, we decided to go with Zig. The main "pros" for Zig were:
+
+- it allows us to call into the Postgres APIs almost directly, so we have the same power as using C
+- it offers more memory safety than C, it's a bit more expressive, and a bit more fun (we think)
```

Comment:
More to add:

## Stack traces

better memory safety and stack traces for a few classes of undefined behavior: https://ziglang.org/documentation/0.11.0/#Undefined-Behavior

My favourite has been me getting a stack trace for forgetting to append a '\0' to a supposedly 0-terminated string.

But in comparison to (safe-)Rust I would say Zig is lacking on the "safe" side.

In general stack traces can be a great time saver.

## Memory/Error handling integration

Especially the ability to integrate with Postgres memory management was a big reason for using Zig

## Typing

Zig comptime helps us to provide type safe wrappers if we see a need (instead of juggling with void pointers).

Reply:
Added another line item.